File: 0437-path-sum-iii/0437-path-sum-iii.py
class Solution:
    def pathSum(self, root: Optional[TreeNode], targetSum: int) -> int:
        def dfs(node, pathSum):
            if not node:
                return 
            currentSum = pathSum + node.val
            x = currentSum - targetSum
            if x in freq:
                self.count += freq[x]
            if currentSum in freq:
                freq[currentSum] += 1
            else:
                freq[currentSum]=1
            dfs(node.left,currentSum)
            dfs(node.right,currentSum)
            freq[currentSum] -= 1
        self.count = 0
        freq = {0:1}
        dfs(root,0)
        return self.count

File: 0437-path-sum-iii/0437-path-sum-iii.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int count = 0;
    int pathSum(TreeNode* root, int targetSum) {
        unordered_map<long,int> hash;
        hash[0] = 1;
        recCount(root,hash,targetSum,0);
        return count;

    }
    void recCount(TreeNode* root, unordered_map<long,int>& hash,int targetSum, long pathSum )
    {
        if(!root)
            return;
        
        long currentSum = pathSum + root->val;
        long x = currentSum - targetSum;

        if(hash.find(x) != hash.end())
            count+= hash[x];
        hash[currentSum]++;
        if(root->left)
            recCount(root->left,hash,targetSum,currentSum);
        if(root->right)
            recCount(root->right,hash,targetSum,currentSum);
        hash[currentSum]--;        
    }
};